app/autotranslate/server/msTranslate.js

```diff
@@ -0,0 +1,174 @@
+		if (this.autoTranslateEnabled && this.apiKey) {
+			if (this.supportedLanguages[target]) {
+				return this.supportedLanguages[target];
+			}
+			const languages = HTTP.get(this.apiGetLanguages);
+			this.supportedLanguages[target] = Object.keys(languages.data.translation).map((language) => ({
+				language,
+				name: languages.data.translation[language].name,
+			}));
+			return this.supportedLanguages[target || 'en'];
+		}
+	}
+
+	/**
+	 * Send Request REST API call to the service provider.
+	 * Returns translated message for each target language in target languages.
+	 * @private
+	 * @param {object} message
+	 * @param {object} targetLanguages
+	 * @returns {object} translations: Translated messages for each language
+	 */
+	_translateMessage(message, targetLanguages) {
+		let translations = {};
+		let msgs = message.msg.split('\n');
+		msgs = msgs.map((msg) => ({ Text: msg }));
+		const supportedLanguages = this.getSupportedLanguages('en');
+		targetLanguages = targetLanguages.map((language) => {
+			if (language.indexOf('-') !== -1 && !_.findWhere(supportedLanguages, { language })) {
+				language = language.substr(0, 2);
+			}
+			return language;
+		});
+		const url = `${ this.apiEndPointUrl }&to=${ targetLanguages.join('&to=') }`;
+		try {
+			const result = HTTP.post(url, {
+				headers: {
+					'Ocp-Apim-Subscription-Key': this.apiKey,
+					'Content-Type': 'application/json; charset=UTF-8',
+				},
+				data: msgs,
+			});
+
+			if (result.statusCode === 200 && result.data && result.data.length > 0) {
+				// store translation only when the source and target language are different.
+				// multiple lines might contain different languages => Mix the text between source and detected target if neccessary
+				translations = Object.assign({}, ...targetLanguages.map((language) =>
+					({
+						[language]: result.data.map((line) => line.translations.find((translation) => translation.to === language).text).join('\n'),
+					}),
+				));
+			}
+		} catch (e) {
+			logger.microsoft.error('Error translating message', e);
+		}
+		return translations;
+	}
+
+	/**
+	 * Returns translated message attachment description in target languages.
+	 * @private
+	 * @param {object} attachment
+	 * @param {object} targetLanguages
+	 * @returns {object} translated messages for each target language
+	 */
+	_translateAttachmentDescriptions(attachment, targetLanguages) {
+		let translations = {};
+		const supportedLanguages = this.getSupportedLanguages('en');
+		targetLanguages = targetLanguages.map((language) => {
+			if (language.indexOf('-') !== -1 && !_.findWhere(supportedLanguages, { language })) {
+				language = language.substr(0, 2);
+			}
+			return language;
+		});
+		const url = `${ this.apiEndPointUrl }&to=${ targetLanguages.join('&to=') }`;
+		try {
+			const result = HTTP.post(url, {
+				headers: {
+					'Ocp-Apim-Subscription-Key': this.apiKey,
+					'Content-Type': 'application/json; charset=UTF-8',
+				},
```

Comment:
This part could be extracted to another method since it's duplicated on `_translateMessage`

Reply:
too true. 
Extracted to `_translate`